ddsrecorder_participants/src/cpp/mcap/McapHandler.cpp

```diff
@@ -0,0 +1,460 @@
+{
+    logInfo(DDSRECORDER_MCAP_HANDLER, "Destroying handler.");
+
+    if (paused_)
+    {
+        stop_event_thread_();
+    }
+    else
+    {
+        dump_data_();
+    }
+    mcap_writer_.close();
+}
+
+void McapHandler::add_schema(const fastrtps::types::DynamicType_ptr& dynamic_type)
+{
+    assert(nullptr != dynamic_type);
+    std::string type_name = dynamic_type->get_name();
+    {
+        // Check if it exists already
+        // NOTE: must be unique mutex taken because it could write afterwards
+        std::unique_lock<SchemaMapType> lock(schemas_);
+        if (schemas_.find(type_name) != schemas_.end())
+        {
+            return;
+        }
+
+        // Schema not found, generate from dynamic type and store
+        std::string schema_text = generate_ros2_schema(dynamic_type);
+
+        // TODO remove
+        logInfo(DDSRECORDER_MCAP_HANDLER, "\nAdding schema with name " << type_name << " :\n" << schema_text << "\n");
+
+        // Create schema and add it to writer and to schemas map
+        mcap::Schema new_schema(type_name, "ros2msg", schema_text);
+
+        {
+            std::lock_guard<std::mutex> guard(write_mtx_);
+            mcap_writer_.addSchema(new_schema);
+        }
+
+        schemas_.insert({type_name, std::move(new_schema)});
+    }
+
+    logInfo(DDSRECORDER_MCAP_HANDLER, "Schema created: " << type_name << ".");
+
+    std::lock_guard<PendingSamplesMapType> lock(pending_samples_);
+
+    auto it = pending_samples_.find(type_name);
+    if (it != pending_samples_.end())
+    {
+        add_pending_samples_(type_name);
+        pending_samples_.erase(it);
+    }
+}
+
+void McapHandler::add_data(
+        const DdsTopic& topic,
+        RtpsPayloadData& data)
+{
+    // Check if channel exists
+    // NOTE: must be unique mutex taken because it could write afterwards
+
+    // Add data to channel
+    Message msg;
+    msg.sequence = unique_sequence_number_++;
+    msg.logTime = now();
+    msg.publishTime = fastdds_timestamp_to_mcap_timestamp(data.source_timestamp);
+    msg.dataSize = data.payload.length;
+
+    if (data.payload.length > 0)
+    {
+        auto payload_owner =
+            const_cast<eprosima::fastrtps::rtps::IPayloadPool*>((eprosima::fastrtps::rtps::IPayloadPool*)data.payload_owner);
+
+        if (payload_owner)
+        {
+            payload_pool_->get_payload(
+                data.payload,
+                payload_owner,
+                msg.payload);
+
+            msg.payload_owner = payload_pool_.get();
+            msg.data = reinterpret_cast<std::byte*>(msg.payload.data);
+        }
+        else
+        {
+            logWarning(
+                DDSRECORDER_MCAP_HANDLER,
+                "Payload owner not found in data received.");
+        }
+    }
+    else
+    {
+        logWarning(
+            DDSRECORDER_MCAP_HANDLER,
+            "Received sample with no payload.");
+    }
+
+    try
+    {
+        auto channel_id = get_channel_id_(topic);
```

Comment:
Would not be better to return a bool instead of handling a exception?
A exception should be use when error, not as a expected return of a function https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#e3-use-exceptions-for-error-handling-only

Reply:
You are right, but this way it felt more comfortable to me since the call that could fail is two levels "deep" (get_channel_id->create_channel_id->get_schema_id) . Also, mcap's `addChannel` does not seem to throw any exception at the moment, but it is not `noexcept` so could change in the future.